### ISSUE TITLE
Using hibernate Validator 4.2 in Payara 4.1.1.163

### ISSUE THREAD

**Issue by maheshmarri** [payara/Payara#1231]
# Description #
----------

Using hibernate Validator 4.2 in Payara 4.1.1.163

## Expected Outcome

Product uses hibernate validator 4.2.0 and it seems payara comes with 5.1.2 .How do we override the default one from 5.1.2 to 4.2.0.

## Current Outcome

Right now it the product fails with validation issues because of the conflicts from payara and product uses hibernate validator 4.2.0.

Can you please let me know how can override the current hibernate validator 5.1.2 .

I tried placing the hibernate-validator from _domain/lib/applibs_ folder. It seems getting conflicted withe the one from payara.

## Environment ##

- **Payara Version**: 4.1.1.163
- **Edition**: 
- **JDK Version**: 
- **Operating System**:

**Comment by fturizo** [comment]
@maheshmarri, you can set the `classloading-delegate` option at **WAR** or **EAR** assemblies using the appropriate deployment descriptors to let the server use the libraries bundled in the application instead of the ones in the server. Check the official documentation on [classloading](https://payara.gitbooks.io/payara-server/content/documentation/extended-documentation/classloading.html) for more info.

May I ask why are you downgrading the version of `hibernate-validator` in this case? If you are migrating an application from a previous version I'd recommend using the version included with Payara.

**Comment by maheshmarri** [comment]
Thanks for the response @fturizo ,we do have the classloading-delegate option at WAR in the deployment descriptior,it seems some how it still uses hibernate-validator and fails with validation issues.

One of the reason for us to go with older version of the hibernate-validator ,supporting multiple application servers -websphere and wildfly

**Comment by lprimak** [comment]
can you paste your glassfish-web.xml please?

**Comment by maheshmarri** [comment]
<!DOCTYPE glassfish-web-app PUBLIC "-//GlassFish.org//DTD 
GlassFish Application Server 3.1 Servlet 3.0//EN" 
"http://glassfish.org/dtds/glassfish-web-app_3_0-1.dtd">
<glassfish-web-app>
    <context-root>/</context-root>
    <session-config>
      <session-timeout>30</session-timeout> 
    </session-config>  
    <resource-ref>
	    <res-ref-name>jdbc/xXXXX</res-ref-name>
	    <res-type>javax.sql.DataSource</res-type>
		<jndi-name>java:/jdbc/XXXXX</jndi-name>
	</resource-ref>	
	<class-loader delegate="false"/>
</glassfish-web-app>

**Comment by maheshmarri** [comment]
Can you please provide some inputs ...Is it possible to override the existing hibernate-validator ,that comes with payara 4.1.1.163

**Comment by fturizo** [comment]
@maheshmarri, from the configuration you posted it seems the classloading delegation is disabled. Is the **hibernate-validator** JAR file in the _WEB-INF/lib_ directory of the **WAR**? If that doesn't work you can set the `fish.payara.classloading.delegate` System property to false at domain level and test it again. Let us know if this works.

**Comment by lprimak** [comment]
Also, you can try putting your hibernate validator into <domain_dir>/lib as well

**Comment by maheshmarri** [comment]
Hi I have already tried putting the hibernate validator in <domain_dir>/lib as well ,it seems to be not overriding the once comes with payara.

**Comment by lprimak** [comment]
The only thing I can think of is that other parts of Payara are using hibernate validator and they are loading it from the modules/ directory even before your app is deployed, thus classes are already loaded and cannot be overridden

**Comment by fturizo** [comment]
Hello @maheshmarri, I did a small test to see if the library could be overwritten by adding the hibernate-4.2.0.Final.jar file to the _<domain_dir>/lib_ and setting the `fish.payara.classloading.delegate` JVM option to false. This worked, but only after clearing the **OSGI** Cache of the domain. However, the server initialized with several errors when bootstrapping the Bean Validation API:

```
MultiException stack 1 of 1
java.lang.AbstractMethodError: org.hibernate.validator.engine.ConfigurationImpl.getBootstrapConfiguration()Ljavax/validation/BootstrapConfiguration;
	at org.glassfish.jersey.server.validation.internal.ValidationBinder$ConfiguredValidatorProvider.getTraversableResolver(ValidationBinder.java:322)
	at org.glassfish.jersey.server.validation.internal.ValidationBinder$ConfiguredValidatorProvider.getDefaultValidatorContext(ValidationBinder.java:303)
...
```

In this case, since this is a previous version of the Hibernate Validator implementation, it's causing issues when bootstrapping other components of Payara as @lprimak said, so I don't recommend going with this option. You would have to replace the _<payara>/glassfish/modules/bean-validation.jar_ with a proper module that uses the older version of the Hibernate Validator classes and test it to see if it works.

**Comment by lprimak** [comment]
Fabio, thanks for trying this.

@maheshmarri what is happening is that Payara-packaged Jersey depends on HV 5.x, and who knows what other dependencies depend on it as well.
One of the advantages of using Java EE is that you don't have to "assemble" 100s of dependencies, and use the ones packaged by the app server because they are known to work together.

I still don't understand why you want HV 4.x, as 5.x is backwards compatible.

**Comment by mulderbaba** [comment]
Also HK2 w/ v2.4.0, the DI framework used in Payara, currently bundles hibernate-validator in one of its artefacts by literally copying its classes to overcome OSGI problems. downgrading would introduce more problems IMO so it'd be better if you favour 5.X in your code. @lprimak backward compatibility could be a problem since 4.2.1 demands banvalidation-api with version 1.0.0.GA but 5.1.2 demands beanvalidation-api with version 1.1.0.Final.

**Comment by fturizo** [comment]
Hi @maheshmarri. Were the provided explanations of help to you? How are you doing with this issue?

**Comment by mikecroft** [comment]
Closing as `wontfix` since the later version of hibernate validator is too tightly coupled to fundamental parts of Payara Server to be easily overcome. We could potentially look at addressing this as part of a support contract, but it's not feasible to address this in a general case.
